### PR TITLE
[IMP] payment(_*): show incompatible payment methods/providers in debug

### PR DESCRIPTION
**Pull request by vchu-odoo**
Since 17.0 payment methods were introduced to the user and it became
more difficult to debug them in the payment form. For that reason with
this commit we added the availability report in payment form that can
be displayed in debug mode. The report includes information about
compatibility of each payment method and each provider for the given
payment context and the reason if it is incompatible.

task-2882668

The cleanup is included in this PR as well. 
- Unified variable name
- added a default payment method code to wire transfer

See related PR:

- https://github.com/odoo/enterprise/pull/55409/
-  https://github.com/odoo/upgrade/pull/5858

### CODE REVIEW

File: addons/payment/models/payment_method.py [hunk 1]
```diff
@@ -201,44 +203,87 @@ def _get_compatible_payment_methods(
         :param bool force_tokenization: Whether only payment methods supporting tokenization can be
                                         matched.
         :param bool is_express_checkout: Whether the payment is made through express checkout.
+        :param dict report: Availability report. To get the full report get_compatible_providers
+                            should be called first.In the end of the function execution report will
```
Comment (Demesmaeker) [inline addons/payment/models/payment_method.py:207]:
```suggestion
                            should be called first. In the end of the function execution report will
```

File: addons/payment/models/payment_method.py [hunk 2]
```diff
@@ -201,44 +203,87 @@ def _get_compatible_payment_methods(
...
+                            show all payment methods and their compatibility to the given
+                            payment context and reason if they are incompatible.
         :return: The compatible payment methods.
         :rtype: payment.method
         """
-        # Compute the base domain for compatible payment methods.
-        domain = [('provider_ids', 'in', provider_ids), ('is_primary', '=', True)]
+
```
Comment (Demesmaeker) [inline addons/payment/models/payment_method.py:213]:
```suggestion
```

**Review by Demesmaeker — CHANGES_REQUESTED** [review]
Hello :duck: 

Didn't test it yet, but from the code, I can tell it doesn't follow the spec. Here, in the report, you'll get the first reason why the payment_method isn't available, not a list of all the reasons. =]

File: addons/payment/models/payment_provider.py [hunk 3]
```diff
@@ -487,58 +492,84 @@ def _get_compatible_providers(
...
-                domain, [
-                    '|',
-                    ('available_country_ids', '=', False),
-                    ('available_country_ids', 'in', [partner.country_id.id]),
-                ]
-            ])
+            unfiltered_providers = providers
+            providers = providers.filtered(
+                lambda p: not p.available_country_ids or
+                           partner.country_id.id in p.available_country_ids.ids
```
Comment (Demesmaeker) [inline addons/payment/models/payment_provider.py:506]:
```suggestion
                lambda p: not p.available_country_ids or
                          partner.country_id.id in p.available_country_ids.ids
```

File: addons/payment/models/payment_provider.py [hunk 4]
```diff
@@ -487,58 +492,84 @@ def _get_compatible_providers(
...
-            domain = expression.AND([
-                domain, [
-                    '|',
-                    ('available_currency_ids', '=', False),
-                    ('available_currency_ids', 'in', [currency.id]),
-                ]
-            ])
+            unfiltered_providers = providers
+            providers = providers.filtered(
+                lambda p: not p.available_currency_ids or currency.id in p.available_currency_ids.ids
```
Comment (Demesmaeker) [inline addons/payment/models/payment_provider.py:538]: 100 chars max =]

File: addons/payment/models/payment_provider.py (hunk 3 above)
Comment (Demesmaeker) [reply addons/payment/models/payment_provider.py:506]: In fact:
```suggestion
                lambda p: not p.available_country_ids 
                          or partner.country_id.id in p.available_country_ids.ids
```

**Review by Demesmaeker — CHANGES_REQUESTED** [review]
Hello :duck: 

First pass here, it's a nice feature, but ftr, I still think this task shouldn't skip the testing stage, as even if it's for the debug mode only, it can still use some testing to see if the report is shown as expected. Testers will be able to see a difference from before AND test the many different possibilities of error shown in the report.

So, back to the review ^^

Could you please reword the PR and commits name so they:
- show this doesn't just impact the `payment` module
- get rid of the typo
- ensure it isn't too long

Mostly nitpicking otherwise, good job!

File: addons/payment/models/payment_provider.py (hunk 3 above)
Comment (vchu-odoo) [reply addons/payment/models/payment_provider.py:506]: You suggested to reformat the line only for this case but not for the others (like L523 the same file or in the `payment_method.py` L251). Isn't it better to stick to the same format everywhere? Personally I find it clearer like it is, but if there is a coding style that we follow, I will change it everywhere then :)

Comment (Demesmaeker) [reply addons/payment/models/payment_provider.py:506]: I did it to not make many comments with the same. But it's like when you are doing maths: which one is clearer to you?
```
5,155 +
3 -
119
```
or 
```
5,155
+ 3
- 119
```

File: addons/payment_custom/data/payment_method_data.xml [hunk 5]
```diff
@@ -5,6 +5,7 @@
         <field name="name">Wire Transfer</field>
         <field name="code">wire_transfer</field>
         <field name="sequence">1000</field>
+        <field name="active">False</field>
```
[the remark this answers is not in the document]
Comment (vchu-odoo) [reply addons/payment_custom/data/payment_method_data.xml:8]: In this PR I included a bit of cleanup as you can see in the separate commit. We forgot to add `active=false` to not activate the payment method `wire_transfer` by default without activating the corresponding provider first.

File: addons/payment/models/payment_provider.py (hunk 3 above)
Comment (vchu-odoo) [reply addons/payment/models/payment_provider.py:506]: It depends on the person I guess. But if you insist I will change ofc.

File: addons/payment_demo/const.py [hunk 6]
```diff
@@ -1,6 +1,6 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 # The codes of the payment methods to activate when Demo is activated.
-DEFAULT_PAYMENT_METHOD_CODES = [
+DEFAULT_PAYMENT_METHODS_CODES = [
```
[the remark this answers is not in the document]
Comment (vchu-odoo) [reply addons/payment_demo/const.py:4]: though, the difference is not so big in terms of performance as  long as the list is short which is the case. But I agree that for our purposes set is more appropriate. Thanks

File: addons/payment_custom/models/payment_provider.py [hunk 7]
```diff
@@ -74,3 +74,10 @@ def _transfer_ensure_pending_msg_is_set(self):
         )
         if transfer_providers_without_msg:
             transfer_providers_without_msg.action_recompute_pending_msg()
+
+    def _get_default_payment_method_codes(self):
```
[the remark this answers is not in the document]
Comment (vchu-odoo) [reply addons/payment_custom/models/payment_provider.py:79]: as was agreed with @AntoineVDV the name of the constant will be rather changed to `DEFAULT_PAYMENT_METHOD_CODES` to correspond to the name of the function.

File: addons/payment_demo/const.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Demesmaeker) [reply addons/payment_demo/const.py]: Not big at all for these kinds of things, and had that be the only comment I had to make, I wouldn't have. It's almost just a personal preference, but since it's something I try to use more often recently, I thought you would be interested too ^^

File: addons/payment/models/payment_provider.py [hunk 8]
```diff
@@ -487,58 +493,85 @@ def _get_compatible_providers(
...
-                    '|', '|',
-                    ('maximum_amount', '>=', converted_amount),
-                    ('maximum_amount', '=', False),
-                    ('maximum_amount', '=', 0.),
-                ]
-            ])
+            unfiltered_providers = providers
+            providers = providers.filtered(
+                lambda p: not p.maximum_amount
+                          or p.maximum_amount >= converted_amount
```
Comment (Demesmaeker) [inline addons/payment/models/payment_provider.py:525]: Hm, might be interesting to use `float_compare` here, to avoid (very rare, I know) problems in the future? :thinking:  Not sure though, should be good like this too.

File: addons/payment_alipay/models/payment_provider.py [hunk 9]
```diff
@@ -33,16 +34,25 @@ class PaymentProvider(models.Model):
...
         if currency and currency.name != 'CNY':
+            unfiltered_providers = providers
             providers = providers.filtered(
                 lambda p: p.code != 'alipay' or p.alipay_payment_method != 'express_checkout'
             )
+            payment_utils.add_to_report(
+                report,
+                unfiltered_providers - providers,
+                available=False,
+                reason=_("incompatible currency provider, only CNY is supported")
```
Comment (Demesmaeker) [inline addons/payment_alipay/models/payment_provider.py:54]:
```suggestion
                reason=_("incompatible currency provider, only CNY is supported"),
```

File: addons/payment_custom/models/payment_provider.py [hunk 10]
```diff
@@ -2,6 +2,7 @@
 
 from odoo import _, api, fields, models
 
+from odoo.addons.payment_custom import const
```
Comment (Demesmaeker) [inline addons/payment_custom/models/payment_provider.py:5]:
```suggestion
from odoo.addons.payment_custom import const

```

File: addons/payment_ogone/models/payment_provider.py [hunk 11]
```diff
@@ -48,12 +50,23 @@ def _compute_feature_support_fields(self):
...
     @api.model
-    def _get_compatible_providers(self, *args, is_validation=False, **kwargs):
+    def _get_compatible_providers(self, *args, is_validation=False, report=None, **kwargs):
         """ Override of payment to unlist Ogone providers for validation operations. """
-        providers = super()._get_compatible_providers(*args, is_validation=is_validation, **kwargs)
+        providers = super()._get_compatible_providers(
+            *args,
+            is_validation=is_validation,
+            report=report,
+            **kwargs)
```
Comment (Demesmaeker) [inline addons/payment_ogone/models/payment_provider.py:59]:
```suggestion
            **kwargs,
        )
```

File: addons/sale/controllers/portal.py [hunk 12]
```diff
@@ -206,19 +206,22 @@ def _get_payment_values(self, order_sudo, downpayment=False, **kwargs):
             amount = order_sudo.amount_total - order_sudo.amount_paid
         currency = order_sudo.currency_id
 
+        report = {}  # Availability report.
```
Comment (Demesmaeker) [inline addons/sale/controllers/portal.py:209]: `availibility_report` then to stay consistent ^^

File: addons/website_payment/models/payment_provider.py [hunk 13]
```diff
@@ -3,8 +3,10 @@
 from werkzeug.urls import iri_to_uri
 
 from odoo import api, fields, models
-from odoo.http import request
+from odoo.addons.payment import utils as payment_utils
+from odoo.addons.payment.const import REPORT_REASONS_MAP
 
+from odoo.http import request
```
Comment (Demesmaeker) [inline addons/website_payment/models/payment_provider.py:9]:
```suggestion
from odoo.http import request

from odoo.addons.payment import utils as payment_utils
from odoo.addons.payment.const import REPORT_REASONS_MAP
```

File: addons/website_payment/models/payment_provider.py [hunk 14]
```diff
@@ -16,21 +18,34 @@ class PaymentProvider(models.Model):
...
+        :param dict report: Availability report.
         :return: The compatible providers.
         :rtype: payment.provider
         """
-        providers = super()._get_compatible_providers(*args, website_id=website_id, **kwargs)
+        providers = super()._get_compatible_providers(
+            *args,
+            website_id=website_id,
+            report=report,
+            **kwargs
```
Comment (Demesmaeker) [inline addons/website_payment/models/payment_provider.py:36]:
```suggestion
            **kwargs,
```

File: addons/payment/views/payment_form_templates.xml [hunk 15]
```diff
@@ -133,9 +145,75 @@
                 <!-- === Submit button === -->
                 <t t-if="display_submit_button" t-call="payment.submit_button"/>
             </div>
+            <!-- === Debug availability report === -->
+            <t t-if="report" groups="base.group_no_one" t-call="payment.availability_report"/>
```
Comment (Demesmaeker) [inline addons/payment/views/payment_form_templates.xml:149]:
```suggestion
            <!-- === Availability report (for debug) === -->
            <t t-if="report" groups="base.group_no_one" t-call="payment.availability_report"/>
```

File: addons/payment/views/payment_form_templates.xml [hunk 16]
```diff
@@ -101,6 +101,18 @@
                     >
                         <t t-if="not collapse_payment_methods">Choose a payment method</t>
                         <t t-else="">Other payment methods</t>
+                        <!-- === Debug availability report button === -->
```
Comment (Demesmaeker) [inline addons/payment/views/payment_form_templates.xml:104]: same

File: addons/payment/tests/test_payment_method.py [hunk 17]
```diff
@@ -142,3 +143,28 @@ def test_payment_method_not_compatible_with_express_checkout(self):
             self.provider.ids, self.partner.id, is_express_checkout=True
         )
         self.assertNotIn(self.payment_method, compatible_payment_methods)
+
+    def test_payment_method_availability_report(self):
+        """ Test that report values are correctly set when the payment method is available and
+        unavailable."""
```
Comment (Demesmaeker) [inline addons/payment/tests/test_payment_method.py:149]: Test that report values are correctly set when the payment method is available and when it is unavailable.

**Review by Demesmaeker — CHANGES_REQUESTED** [review]
Hello :duck: 

Mostly nitpicking comments ^^

Regarding the testing, this only covers the beginning, only the case when the currency changed, meaning that the rest of the code isn't tested. I can see 2 ways of doing it:
- Real, full unittests. So, a lot of tests, one for each case. _(Personal recommendation, as it'll then also make sense to test the specific cases in the same way for subscription, on site picking, website_sale, Alipay... + they'll be nearly identical so really fast to do.)_
- 2 big tests (1 for the providers, 1 for the PM), where you set every wrong condition possible. Faster on the runbot (kinda), but hell to debug/read... Not exactly what we usually do.
Now, this depends on whether we want a full test coverage on this feature. IMHO, it is interesting to do it because since it's a debug feature, it means that it should be used less often than normal features, but when it is, having misinformation there might mislead our support team (and then us) for a while, and so, it'd take a lot of time before we even reckon there is a problem in the report.

**Comment by vchu-odoo** [comment]
Thanks for the review, @Demesmaeker!
Regarding the tests, there are already tests called like `test_payment_method_compatible_*` for pms for ex., that already check if the filtering works correctly. The only difference between those and what you suggest is the check of values in the report which are put there via the same function `add_to_report`. So maybe it is better to include testing report values inside already existing functions or add tests for `add_to_report` ?

File: addons/payment/models/payment_provider.py (hunk 8 above)
Comment (vchu-odoo) [reply addons/payment/models/payment_provider.py:525]: sure, but which rounding should we take?

**Comment by Demesmaeker** [comment]
> So maybe it is better to include testing report values inside already existing functions or add tests for `add_to_report` ?

Yes, indeed, sorry, I forgot about these tests :sweat_smile:  . Checking it there would indeed be perfect IMHO. =]

File: addons/payment/models/payment_method.py [hunk 18]
```diff
@@ -4,6 +4,8 @@
 from odoo.exceptions import UserError
 from odoo.osv import expression
 
+from odoo.addons.payment import utils as payment_utils
+from odoo.addons.payment.const import REPORT_REASONS_MAP
```
Comment (Demesmaeker) [inline addons/payment/models/payment_method.py:8]: missing second new line before the class

File: addons/payment/tests/test_payment_method.py [hunk 19]
```diff
@@ -142,3 +143,28 @@ def test_payment_method_not_compatible_with_express_checkout(self):
             self.provider.ids, self.partner.id, is_express_checkout=True
         )
         self.assertNotIn(self.payment_method, compatible_payment_methods)
+
+    def test_payment_method_availability_report(self):
+        """ Test that report values are correctly set when the payment method is available and
+        when it is unavailable."""
```
Comment (Demesmaeker) [inline addons/payment/tests/test_payment_method.py:149]:
```suggestion
        """ Test that report values are correctly set when the payment method is available and when 
        it is unavailable."""
```
You don't have to break sooner than 100 chars =]

File: addons/payment/tests/test_payment_provider.py [hunk 20]
```diff
@@ -226,3 +226,32 @@ def test_provider_not_compatible_with_express_checkout(self):
...
+            self.company.id, self.partner.id, self.amount, report=report
+        )
+
+        # Test that report is configured correctly for available payment provider.
+        self.assertIn('providers', report)
+        self.assertIn(self.provider, report['providers'])
+        self.assertTrue(report['providers'][self.provider]['available'])
+        self.assertFalse(report['providers'][self.provider]['reason'])
+
+        # Test that for incompatible provider 'available' is set to False and the reason is defined.
```
Comment (Demesmaeker) [inline addons/payment/tests/test_payment_provider.py:245]:
```suggestion
        # Test that for an incompatible provider, 'available' is set to False and the reason is defined.
```

File: addons/payment/tests/test_payment_method.py [hunk 21]
```diff
@@ -142,3 +143,28 @@ def test_payment_method_not_compatible_with_express_checkout(self):
...
+
+        report = {}
+        payment_utils.add_to_report(report, self.provider)  # add available provider
+        self.env['payment.method']._get_compatible_payment_methods(
+            self.provider.ids, self.partner.id, report=report
+        )
+        # Check that report is configured correctly for available payment method.
```
Comment (Demesmaeker) [inline addons/payment/tests/test_payment_method.py:156]:
```suggestion
        # Check that the report is configured correctly for available payment methods.
```

File: addons/payment/tests/test_payment_method.py [hunk 22]
```diff
@@ -142,3 +143,28 @@ def test_payment_method_not_compatible_with_express_checkout(self):
...
+        self.assertIn('payment_methods', report)
+        self.assertIn(self.payment_method, report['payment_methods'])
+        self.assertTrue(report['payment_methods'][self.payment_method]['available'])
+        self.assertIn('supported_providers', report['payment_methods'][self.payment_method])
+
+        # Check that for incompatible pm 'available' is set to False and the reason is defined.
```
Comment (Demesmaeker) [inline addons/payment/tests/test_payment_method.py:162]:
```suggestion
        # Check that for an incompatible pm, 'available' is set to False and the reason is defined.
```

File: addons/payment/tests/test_payment_provider.py [hunk 23]
```diff
@@ -226,3 +226,32 @@ def test_provider_not_compatible_with_express_checkout(self):
...
+    def test_payment_provider_availability_report(self):
+        """ Test that report values are correctly set when the provider is available and
+        unavailable. """
+
+        report = {}
+        self.env['payment.provider']._get_compatible_providers(
+            self.company.id, self.partner.id, self.amount, report=report
+        )
+
+        # Test that report is configured correctly for available payment provider.
```
Comment (Demesmaeker) [inline addons/payment/tests/test_payment_provider.py:239]:
```suggestion
        # Test that the report is configured correctly for an available payment provider.
```

File: addons/payment/utils.py [hunk 24]
```diff
@@ -42,6 +42,50 @@ def check_access_token(access_token, *values):
...
+                    'available': true|false,
+                    'reason': '',
+                    'supported_providers: [(provider_record, report['providers'][p]['available'])],
+                },
+            }
+        }
+    :param dict report: availability report for providers and payment methods
+    :param payment.provider|payment.method records: records to include in the report
+    :param bool is_provider: True if records are of 'payment.provider' model else of
+                            'payment.method'
```
Comment (Demesmaeker) [inline addons/payment/utils.py:69]: Could you reformulate that?

File: addons/payment/utils.py [hunk 25]
```diff
@@ -42,6 +42,50 @@ def check_access_token(access_token, *values):
...
+    :param bool available: True if record is available else False
```
Comment (Demesmaeker) [inline addons/payment/utils.py:70]:
```suggestion
    :param bool available: Whether the record is available or not.
```

File: addons/payment/utils.py [hunk 26]
```diff
@@ -42,6 +42,50 @@ def check_access_token(access_token, *values):
...
+
+def add_to_report(report, records, is_provider=True, available=True, reason=''):
+    """ Add records to the report with the provided values.
+
+        Structure of the report:
+        report = {
+            'providers': {
+                provider_record : {
+                    'available': true|false,
+                    'reason': ''
```
Comment (Demesmaeker) [inline addons/payment/utils.py:55]:
```suggestion
                    'reason': "",
```

File: addons/payment/utils.py [hunk 27]
```diff
@@ -42,6 +42,50 @@ def check_access_token(access_token, *values):
...
+                },
+            },
+            'payment_methods': {
+                pm_record : {
+                    'available': true|false,
+                    'reason': '',
```
Comment (Demesmaeker) [inline addons/payment/utils.py:61]:
```suggestion
                    'reason': "",
```

File: addons/payment/data/ir_actions_server_data.xml [hunk 28]
```diff
@@ -3,12 +3,10 @@
...
     <record id="action_activate_stripe" model="ir.actions.server">
         <field name="name">Activate Stripe</field>
-        <field name="model_id" ref="website_payment.model_payment_provider"/>
+        <field name="model_id" ref="payment.model_payment_provider"/>
         <field name="state">code</field>
         <field name="code">
-menu = env.ref('website.menu_website_website_settings', raise_if_not_found=False)
-menu_id = menu and menu.id
-action = env.company._run_payment_onboarding_step(menu_id=menu_id)
+action = env.company._run_payment_onboarding_step()
```
Comment (Demesmaeker) [inline addons/payment/data/ir_actions_server_data.xml:9]: Add and explain why this change in the task too please

**Review by Demesmaeker — CHANGES_REQUESTED** [review]
Hello :duck: 

Mostly nitpicking but what about the previous comment about the tests? Didn't you tell me that you'd add checks on all the tests for `_get_compatible_payment_methods` and `_get_compatible_providers`?

File: addons/payment/data/ir_actions_server_data.xml (hunk 28 above)
Comment (vchu-odoo) [reply addons/payment/data/ir_actions_server_data.xml:9]: I mentioned it in dev notes and for convenience duplicated it here.
As an action `action_activate_stripe​` should be available for all modules that call `payment.form​` (because we move payment warning to the form) the menu of website setting was removed to allow it to fall back to the default one.

File: addons/payment/tests/test_payment_provider.py (hunk 23 above)
Comment (Demesmaeker) [reply addons/payment/tests/test_payment_provider.py:240]: Still there

File: addons/payment/tests/test_payment_provider.py (hunk 20 above)
Comment (Demesmaeker) [reply addons/payment/tests/test_payment_provider.py:246]: Comment resolved without change

File: addons/payment/tests/test_payment_method.py [hunk 29]
```diff
@@ -142,3 +144,94 @@ def test_payment_method_not_compatible_with_express_checkout(self):
...
+                pm_no_express: {
+                    'available': False,
+                    'reason': REPORT_REASONS_MAP['express_checkout_not_supported'],
+                    'supported_providers': [(self.provider, True)],
+                }
+        }
+        self.assertIn(
+            'providers',
+            report,
+            msg="Method _get_compatible_providers should be run first to get a full report."
```
Comment (Demesmaeker) [inline addons/payment/tests/test_payment_method.py:234]:
```suggestion
            msg="Method _get_compatible_providers should be run first to get a full report.",
```

File: addons/payment/tests/test_payment_provider.py [hunk 30]
```diff
@@ -226,3 +227,112 @@ def test_provider_not_compatible_with_express_checkout(self):
             self.company.id, self.partner.id, self.amount, is_express_checkout=True
         )
         self.assertNotIn(self.provider, compatible_providers)
+
+    def test_payment_provider_availability_report(self):
```
Comment (Demesmaeker) [inline addons/payment/tests/test_payment_provider.py:231]: Seeing the next test, is this one still needed? Or as for the payment provider, only one where everything is tested is enough?

**Review by Demesmaeker — APPROVED** [review]
Hello :duck: 

Thanks for the test, as the comment left are only nitpicking, I'll already green that task, but they still need to be addressed.

Have a nice day,

File: addons/payment/tests/test_payment_provider.py (hunk 20 above)
Comment (vchu-odoo) [reply addons/payment/tests/test_payment_provider.py:246]: Oops, :see_no_evil: , thanks for the remind!

File: addons/payment/tests/test_payment_provider.py (hunk 30 above)
Comment (vchu-odoo) [reply addons/payment/tests/test_payment_provider.py:231]: Yes, you are right, forgot to remove it.

**Review by AntoineVDV — CHANGES_REQUESTED** [review]

File: addons/payment/views/payment_form_templates.xml [hunk 31]
```diff
@@ -133,7 +135,119 @@
...
+    <template id="payment.availability_report">
+        <div
+            t-if="availability_report"
+            class="collapse"
+            id="payment_availability_report"
+            groups="base.group_no_one"
+        >
+            <h4 class="fs-6 text-uppercase fw-bolder">Availability report</h4>
+                <t t-call="payment.availability_report_records">
+                    <t t-set="is_pm" t-value="False"/>
```
[the remark this answers is not in the document]
Comment (vchu-odoo) [reply addons/payment/views/payment_form_templates.xml:196]: records here is dict, so I do not see how it could be better.
`records` name may be a bit confusing, so I can rename it to `record_dict` or smth.

File: addons/website_sale/views/templates.xml [hunk 32]
```diff
@@ -2120,38 +2120,10 @@
...
-                        </div>
-                        <a role="button"
-                            class="btn-link alert-warning ps-2"
-                            t-attf-href="/web#action=#{payment_action_id}"
-                        >
-                            <strong><i class="oi oi-arrow-right"></i> View alternatives</strong>
-                        </a>
-                    </div>
-                </div>
+                <t t-if="not (payment_methods_sudo or tokens_sudo)" t-set="hide_payment_button" t-value="True"/>
```
[the remark this answers is not in the document]
Comment (vchu-odoo) [reply addons/website_sale/views/templates.xml:2126]: it hides payment button when no pms are available. First image is if I remove it.
image
image
